src/reanimated2/animation/spring.ts

```diff
@@ -38,34 +55,43 @@ export function withSpring(
 
   return defineAnimation<SpringAnimation>(toValue, () => {
     'worklet';
-
-    // TODO: figure out why we can't use spread or Object.assign here
-    // when user config is "frozen object" we can't enumerate it (perhaps
-    // something is wrong with the object prototype).
-    const config: Required<SpringConfig> = {
+    const defaultConfig: Record<keyof SpringConfig, any> = {
       damping: 10,
       mass: 1,
       stiffness: 100,
       overshootClamping: false,
       restDisplacementThreshold: 0.01,
       restSpeedThreshold: 2,
       velocity: 0,
-    };
-    if (userConfig) {
-      Object.keys(userConfig).forEach(
-        (key) =>
-          ((config as any)[key] = userConfig[key as keyof typeof userConfig])
-      );
-    }
+      duration: undefined,
+    } as const;
+
+    const config = { ...defaultConfig, ...userConfig };
 
     function spring(animation: InnerSpringAnimation, now: Timestamp): boolean {
-      const { toValue, lastTimestamp, current, velocity } = animation;
+      const { toValue, firstTimestamp, lastTimestamp, current, velocity } =
+        animation;
+
+      const timeFromStart = now - firstTimestamp;
+      if (userConfig?.duration && timeFromStart > userConfig.duration * 1.1) {
+        animation.current = toValue;
+
+        // clear lastTimestamp to avoid using stale value by the next spring animation that starts after this one
+        animation.lastTimestamp = 0;
+        return true;
+      }
 
       const deltaTime = Math.min(now - lastTimestamp, 64);
       animation.lastTimestamp = now;
 
-      const c = config.damping;
-      const m = config.mass;
+      const c =
+        config.duration && !userConfig?.damping
+          ? animation.newDamping
```

Comment:
would it be possible to only make this check when the animation is created instead of doing it in every frame. In that case we'd just use `animation.damping` here instead of having both `config.damping` and `animation.newDamping` and only set that `animation.damping` at the start (it can't really be changed in-flight).

Same applies to mass